### PR TITLE
Re-write the parsing engine

### LINKED ISSUE #189: Single Line Errors
Only tested with C++, but errors are only displayed with one line now. 
For example if a static_assert fails, the only error displayed is this

```
in instantiation of function template specialization 'Entity<EntityManager<CRender, CLocation> >::addComponent<Scene>' requested here
```

where before it would show multiple lines of information about the error and the build output shows

```
/home/nick/Programming/OpenGL-Project/./src/Entity/entity.hpp:172:9: error: static_assert failed "index() called on invalid type"
        static_assert(contains<comp_t>(), "index() called on invalid type");
        ^             ~~~~~~~~~~~~~~~~~~
/home/nick/Programming/OpenGL-Project/./src/Entity/entity.hpp:68:61: note: in instantiation of function template specialization
      'EntityManager<CRender, CLocation>::index<Scene>' requested here
        auto loc = m_componentIDs[m_entityManager->template index<comp_t>()];
                                                            ^
/home/nick/Programming/OpenGL-Project/src/Scene/scene.cpp:49:9: note: in instantiation of function template specialization 'Entity<EntityManager<CRender, CLocation>
      >::addComponent<Scene>' requested here
    ent.addComponent<Scene>();
```

as you can see, it chops of the beginning of the error, while that is the most important

I originally posted this in the linter repo, but the owner said

> The `excerpt` and `description` have been split into two separate fields now, to get rid of the behavior you're experiencing, you should request the linter provider owner to upgrade their package to Linter v2 API

So I'm guessing this error might belong here? It mostly happens when static_asserts fail

### LINKED ISSUE #195: Linter: Error running clang Error since V4.0.0
Hello,

I updated to v4 and and it seems like the plugin can't compile anymore. Each time I save (it try to compile automatically) I've got a strange error message:
```
/Users/Timothy/.atom/packages/linter/lib/linter-registry.js:177 
[Linter] Error running clang Error: Line number (292) greater than maximum line (241)    
    at Object.generateRange (/Users/Timothy/.atom/packages/linter-clang/node_modules/atom-linter/lib/index.js:158:11)
    at Object.<anonymous> (/Users/Timothy/.atom/packages/linter-clang/lib/main.js:190:32)
    at next (<anonymous>)
    at step (/Users/Timothy/.atom/packages/linter-clang/lib/main.js:7:273)
```

Here is a screenshot:
capture d ecran 2017-03-31 a 09 43 36

### PR DESCRIPTION
**Pull request by Arcanemagus**
Complete re-write of the parsing of clang's messages:
* Support suggested fixes!
* Handle multiple ranges from clang, using a merger of all given
* Better support for ranges in other files
* Much more complete specs

Example of the new parser handling clang output: https://regex101.com/r/sqhHfx/4

Fixes #189.
Fixes #195.

### CODE REVIEW

**Comment by Arcanemagus** [comment]
Note! This is about as good as I could get with RegExp, this will _not_ show a message on the file you are editing if there are only issues in a file included in it, so I would recommend setting the panel to `project` for this.

This linter is only running on a single file, so we can't set the provider scope to file as that would clear _all_ messages handled by it, potentially clearing messages that shouldn't be.

File: lib/main.js [hunk 1]
```diff
@@ -168,35 +242,62 @@ export default {
...
+            excerpt = match[6];
+          }
+          const message = {
             severity,
             location: { file, position },
-            excerpt: match[9],
-          });
+            excerpt,
+          };
+          if (match[8] !== undefined) {
```
Comment (steelbrain) [inline lib/main.js]: This could be shortened to `if (match[8])` without any side-effects

File: lib/main.js [hunk 2]
```diff
@@ -168,35 +242,62 @@ export default {
...
+              position = helpers.generateRange(editor, line, col);
+            }
           }
-          const severity = /error/.test(match[8]) ? 'error' : 'warning';
-          // Handle paths other than the file being linted still
-          const file = match[1] === '<stdin>' ? filePath : match[1];
-          toReturn.push({
+          const severity = /error/.test(match[5]) ? 'error' : 'warning';
+          let excerpt;
+          if (match[7] !== undefined) {
```
Comment (steelbrain) [inline lib/main.js]: This could be shortened to `if (match[7])` without any side-effects

File: lib/main.js [hunk 3]
```diff
@@ -168,35 +242,62 @@ export default {
...
         const toReturn = [];
-        let match = regex.exec(output);
 
+        let match = regex.exec(output);
         while (match !== null) {
+          const isCurrentFile = match[1] === '<stdin>';
+          // If the "file" is stdin, override to the current editor's path
+          const file = isCurrentFile ? filePath : match[1];
           let position;
           if (match[4] !== undefined) {
```
Comment (steelbrain) [inline lib/main.js]: Same here, This could be shortened to `if (match[4])` without any side-effects

File: lib/main.js [hunk 4]
```diff
@@ -6,7 +6,80 @@ import { dirname, extname } from 'path';
...
+/**
+ * Determine if a given path is open in an existing TextEditor
+ * @param  {String} filePath The file path to search for an editor of
+ * @return {TextEditor | false}      The TextEditor or false if none found
+ */
+const findTextEditor = (filePath) => {
+  const allEditors = atom.workspace.getTextEditors();
+  const matchingEditor = allEditors.find(
+    textEditor => textEditor.getPath() === filePath);
+  if (matchingEditor !== undefined) {
```
Comment (steelbrain) [inline lib/main.js]: No need to be super-explicit here as well, could just be `return matchingEditor || false` 😛

**Pull request merged**